solver/src/liquidity_collector.rs

```diff
@@ -0,0 +1,38 @@
+use anyhow::{Context, Result};
+
+use crate::{
+    liquidity::uniswap::UniswapLikeLiquidity, liquidity::Liquidity, orderbook::OrderBookApi,
+};
+
+pub struct LiquidityCollector {
+    pub uniswap_like_liquidity: Vec<UniswapLikeLiquidity>,
+    pub orderbook_api: OrderBookApi,
+}
+
+impl LiquidityCollector {
+    pub async fn get_liquidity(&self) -> Result<Vec<Liquidity>> {
+        let limit_orders = self
+            .orderbook_api
+            .get_liquidity()
+            .await
+            .context("failed to get orderbook")?;
+        tracing::debug!("got {} orders", limit_orders.len());
+
+        let mut amms = vec![];
+        for liquidity in self.uniswap_like_liquidity.iter() {
+            amms.extend(
+                liquidity
+                    .get_liquidity(limit_orders.iter())
+                    .await
+                    .context("failed to get pool")?,
+            );
+        }
```

Comment:
I tried to use `flat_map` here, but ran into some issue with the async calls inside. @vkgnosis any tricks up your sleeve that could make this more rustic?

Reply:
I think you might have to convert the iterator into a `Stream`.

Reply:
Thanks I'll give this a try.

Reply:
The [`StreamExt::flat_map` example](https://docs.rs/futures/0.3.14/futures/stream/trait.StreamExt.html#examples-13) is almost exactly what you want.

Reply:
That's even better.

Reply:
Nothing wrong with the current imperative code either. Feel free to use which one you like more.

Reply:
`StreamExt::flat_map` looks nice. The way I read the docs however, it's not actually parallelising the inner stream. 

> The provided closure which produce inner streams is executed over all elements of stream as last inner stream is terminated and next stream item is available.

But it's not super important here and might be easier to read than what we do in the [PoolFetcher](https://github.com/gnosis/oba-services/blob/089d91775bfa64043dddcc6e3355d44cb769b757/shared/src/pool_fetching.rs#L249-L270) to achieve concurrency of the inner futures.

Reply:
I actually personally _ever so slightly_ prefer the current implementation (without `stream::iter`) as I find the `for` loop to be more readable in this case.

Either way, if you prefer using `Stream` and `flat_map` go for it.

Reply:
Nope. I'm gonna leave it as is. Much easier for me to read also.